File: libs/sigc++2/sigc++/connection.cc
// -*- c++ -*-
/*
 * Copyright 2002, The libsigc++ Development Team
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 */

#include <sigc++/connection.h>
using namespace std;

namespace sigc {

connection::connection()
: slot_(0)
{}

connection::connection(const connection& c)
: slot_(c.slot_)
{
  //Let the connection forget about the signal handler when the handler object dies:
  if (slot_)
    slot_->add_destroy_notify_callback(this, &notify);
}

connection::connection(slot_base& sl)
: slot_(&sl)
{
  //Let the connection forget about the signal handler when the handler object dies:
  slot_->add_destroy_notify_callback(this, &notify);
}

connection& connection::operator=(const connection& c)
{
  set_slot(c.slot_);
  return *this;
}

connection::~connection()
{
  if (slot_)
    slot_->remove_destroy_notify_callback(this);
}

bool connection::empty() const
{
  return (!slot_ || slot_->empty());
}

bool connection::connected() const
{
  return !empty();
}

bool connection::blocked() const
{
  return (slot_ ? slot_->blocked() : false);
}

bool connection::block(bool should_block)
{
  return (slot_ ? slot_->block(should_block) : false);
}

bool connection::unblock()
{
  return (slot_ ? slot_->unblock() : false);
}

void connection::disconnect()
{
  if (slot_)
    slot_->disconnect(); // This notifies slot_'s parent.
} 

connection::operator bool()
{
  return !empty();
}
    
void connection::set_slot(slot_base* sl)
{
  if (slot_)
    slot_->remove_destroy_notify_callback(this);

  slot_ = sl;

  if (slot_)
    slot_->add_destroy_notify_callback(this, &notify);
}

void* connection::notify(void* data)
{
  connection* self = reinterpret_cast<connection*>(data);
  self->slot_ = 0;
  return 0;
}

} /* namespace sigc */

File: share/websockets_client/browser/js/connection.js
/*
 * Copyright (C) 2020 Luciano Iam <lucianito@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

const JSON_INF = 1.0e+128;

class Connection {

    constructor (host, port) {
        this.socket = new WebSocket(`ws://${host}:${port}`);
        this.socket.onopen = () => this.openCallback();
        this.socket.onclose = () => this.closeCallback();
        this.socket.onerror = (error) => this.errorCallback(error);
        this.socket.onmessage = (event) => this._onMessage(event);
    }
    
    openCallback () {
        // empty
    }

    closeCallback () {
        // empty
    }
    
    errorCallback (error) {
        // empty
    }

    messageCallback (node, addr, val) {
        // empty
    }

    send (node, addr, val) {
        for (const i in val) {
            if (val[i] == Infinity) {
                val[i] = JSON_INF;
            } else if (val[i] == -Infinity) {
                val[i] = -JSON_INF;
            }
        }
        
        const json = JSON.stringify({node: node, addr: addr, val: val});

        this.socket.send(json);
    }

    _onMessage (event) {
        const msg = JSON.parse(event.data);

        for (const i in msg.val) {
            if (msg.val[i] >= JSON_INF) {
                msg.val[i] = Infinity;
            } else if (msg.val[i] <= -JSON_INF) {
                msg.val[i] = -Infinity;
            }
        }

        this.messageCallback(msg.node, msg.addr || [], msg.val);
    }

}
